README.md
# Instance History Extra
A WoW Classic addon that tracks your recently entered instances and lets you know when you are potentially locked out.

![Preview of addon in action.](https://silverhawke.s-ul.eu/xeLo44RU)

This addon...

* Handles both the "30 instances per day" and the "5 instances per hour" limit.
  * It also handles the 30-minute auto-reset case!
* Shows the duration you spent in an instance as a progress bar. The progress bar represents 24 hours or 1 hour, depending on how many instances you have entered recently and other settings.
  * Generally, you can quickly toggle between 24h/1h view by holding Ctrl while hovering over the display. It isn't foolproof but it works.
* Displays expected time until more instances are available.
* Works in most cases, but most notably it cannot detect if: a reset is done, the user is not the party leader, and the party leader doesn't use this addon (or the WeakAura).
  * In this case, typing `/ihex forcereset` will allow the addon to count the next instance as a new one.

## Installing
It is recommended to download from [CurseForge](https://www.curseforge.com/wow/addons/instancehistoryextra). Alternatively, download or clone from GitHub.

## Configuration
Type `/ihex` or go to `Main Menu → Interface Options → AddOns → Instance History Extra`.

## Acknowledgements
Most of the code was based on the [Instance History](https://wago.io/OXlZupyKm) WeakAura, which was in turn based on the [SavedInstance](https://www.curseforge.com/wow/addons/saved_instances) addon.

## License
Copyright © 2020 silverhawke | [MIT License](https://opensource.org/licenses/MIT)


changelog.txt
v1.0.2:
- Make instance history separate for each character, following July 8's hotfix.

v1.0.1:
- Implement a better handling of the quick toggle feature.

v1.0.0:
- Initial release.


Main.lua
local addonName, env = ...

-- Defaults, and instantiate saved variable
env.configDefaults = {
    debugMode           = false,
    reportResets        = false,
    reportLockedOnly    = false,
    colorProgress       = true,
    force24H            = false,
    updateInterval      = 5,
    displayMin          = 2,
    width               = 300,
    height              = 15,
    xOffset             = 0,
    yOffset             = 55,
    fontSize            = 10,
}
InstanceHistoryExtraSV = InstanceHistoryExtraSV or {}

-- EVENT HANDLERS --
env.onEvent = {}

function env.onEvent.ADDON_LOADED(...)
    local n = select(2, ...)
    if n == addonName then
        -- Instantiate database
        if not InstanceHistoryExtraSV[env.c.thisToon] then
            InstanceHistoryExtraSV[env.c.thisToon] = {}
        end
        local db = InstanceHistoryExtraSV[env.c.thisToon]
        db.histGeneration = db.histGeneration or 1
        db.History = db.History or {}
        db.Instances = db.Instances or {}
        db.colorOffset = db.colorOffset or 0
        local config = InstanceHistoryExtraSV.CONFIG

        -- Fill in missing values
        env.updateTable(config, env.configDefaults, true)

        -- Remove unused configs
        for k, _ in pairs(config) do
            if env.configDefaults[k] == nil then
                config[k] = nil
            end
        end

        -- Session variable
        db.sess = {}
        db.lastDisplayUpdate = 0

        -- Load progress bar saved settings
        env.f.drawProgressBar()

        -- Populate options frame
        env.optionsFrame.populateOptions()

        env.f.updateProgress()
        env.f.updateText()
    end
end

function env.onEvent.CHAT_MSG_SYSTEM(...)
    local msg = select(2, ...)
    local raiddiffmsg = ERR_RAID_DIFFICULTY_CHANGED_S:gsub("%%s",".+")
    local dungdiffmsg = ERR_DUNGEON_DIFFICULTY_CHANGED_S:gsub("%%s",".+")

    if msg == INSTANCE_SAVED then -- just got saved
        C_Timer.After(4, env.f.HistoryUpdate)
    elseif (msg:match("^"..raiddiffmsg.."$") or msg:match("^"..dungdiffmsg.."$")) and
            not env.f.histZoneKey() then -- ignore difficulty messages when creating a party while inside an instance
        env.f.HistoryUpdate(true)
    elseif msg:match(TRANSFER_ABORT_TOO_MANY_INSTANCES) then
        env.f.HistoryUpdate(false,true)
    end
end

function env.onEvent.INSTANCE_BOOT_START()
    env.f.HistoryUpdate(true)
end

function env.onEvent.INSTANCE_BOOT_STOP()
    local db = InstanceHistoryExtraSV[env.c.thisToon]
    if env.f.InGroup() then
        db.sess.delayedReset = false
    end
end

function env.onEvent.GROUP_ROSTER_UPDATE()
    local db = InstanceHistoryExtraSV[env.c.thisToon]
    if db.sess.histInGroup and not env.f.InGroup() and -- ignore failed invites when solo
            not env.f.histZoneKey() then -- left group outside instance, resets now
        env.f.HistoryUpdate(true)
    end
end

function env.onEvent.PLAYER_ENTERING_WORLD()
    C_Timer.After(6, function()
        local db = InstanceHistoryExtraSV[env.c.thisToon]
        db.sess.enterLoc = env.f.getLocation()
    end)

    env.f.zoneChanged()
end

function env.onEvent.ZONE_CHANGED_NEW_AREA()
    env.f.zoneChanged()
end

function env.onEvent.RAID_INSTANCE_WELCOME()
    env.f.zoneChanged()
end

function env.onEvent.PLAYER_CAMPING()
    local db = InstanceHistoryExtraSV[env.c.thisToon]
    db.lastLoc = env.f.getLocation()
end

function env.onEvent.CHAT_MSG_ADDON(...)
    local pre, msg, _, sender = select(2, ...)
    if pre == env.c.prefix then
        if msg == "GENERATION_ADVANCE" and not UnitIsUnit(sender, "player") then
            env.f.HistoryUpdate(true)
        elseif msg == "RESET_REQUEST" then
            if UnitIsGroupLeader("player") then
                env.chatMsg(string.format("Received reset request from %s. All instances will be reset once %s is offline.", sender, sender))
                env.autoReset = string.match(sender, "^[^%-]+")
            end
        end
    end
end

function env.onEvent.SEND_INSTANCE_RESET_REQUEST()
    local channel = env.f.InGroup()

    if channel then
        C_ChatInfo.SendAddonMessage(env.c.prefix, "RESET_REQUEST", channel)
    end
end

function env.handler(self, event, ...)
    env.onEvent[event](self, ...)
end

-- Create main frame
local mainFrame = CreateFrame("Frame")
mainFrame:RegisterEvent("ADDON_LOADED")
mainFrame:RegisterEvent("CHAT_MSG_SYSTEM")
mainFrame:RegisterEvent("INSTANCE_BOOT_START")
mainFrame:RegisterEvent("INSTANCE_BOOT_STOP")
mainFrame:RegisterEvent("GROUP_ROSTER_UPDATE")
mainFrame:RegisterEvent("PLAYER_ENTERING_WORLD")
mainFrame:RegisterEvent("ZONE_CHANGED_NEW_AREA")
mainFrame:RegisterEvent("RAID_INSTANCE_WELCOME")
mainFrame:RegisterEvent("PLAYER_CAMPING")
mainFrame:RegisterEvent("CHAT_MSG_ADDON")
--mainFrame:RegisterEvent("SEND_INSTANCE_RESET_REQUEST")
mainFrame:SetScript("OnEvent", env.handler)
C_ChatInfo.RegisterAddonMessagePrefix(env.c.prefix)

-- Create progress bar frame
local progressBar = CreateFrame("Frame", nil, UIParent)
progressBar.textures = {}
progressBar:SetBackdrop({bgFile = env.c.texture})
progressBar:SetBackdropColor(0.3, 0.3, 0.3)
progressBar.toggle = false
local text = progressBar:CreateFontString(nil, "ARTWORK")
text:SetFont(env.c.font, env.configDefaults.fontSize, "OUTLINE")
text:SetTextColor(0.6, 0.6, 0.6)
progressBar.text = text
env.progressBar = progressBar

-- Secret progress bar feature!!!
progressBar:SetScript("OnEnter", function()
    progressBar:RegisterEvent("MODIFIER_STATE_CHANGED")
    -- In case Ctrl is held before entering
    if IsControlKeyDown() then
        progressBar.toggle = true
        env.f.drawProgressBar()
    end
end)
progressBar:SetScript("OnLeave", function()
    progressBar:UnregisterEvent("MODIFIER_STATE_CHANGED")
    -- When cursor leaves frame, revert back if Ctrl is held
    if IsControlKeyDown() then
        progressBar.toggle = false
        env.f.drawProgressBar()
    end
end)
progressBar:SetScript("OnEvent", function(...)
    -- Ignore if it's not Ctrl key
    local key, value = select(3, ...)
    if not key:find("CTRL") then return end

    progressBar.toggle = value == 1 and true or false
    env.f.drawProgressBar()
end)

-- Handle slash command
SLASH_INSTANCEHISTEX1 = "/ihex"
function SlashCmdList.INSTANCEHISTEX(arg)
    local t = {}
    for s in arg:gmatch("([^ ]+)") do
        table.insert(t, s)
    end

    if t[1] == "forcereset" then
        env.chatMsg("Assuming current instance has been reset...")
        env.f.HistoryUpdate(true)
    else
        InterfaceOptionsFrame_OpenToCategory(env.optionsFrame)
    end
end


Helpers.lua
local _, env = ...

function env.chatMsg(...)
    print("\124cFFFF0000Instance History\124r: " .. string.format(...))
end

function env.debug(...)
    local config = InstanceHistoryExtraSV.CONFIG

    if config.debugMode then
        env.chatMsg(...)
    end
end

function env.deepcopy(orig)
    local orig_type = type(orig)
    local copy
    if orig_type == 'table' then
        copy = {}
        for orig_key, orig_value in next, orig, nil do
            copy[env.deepcopy(orig_key)] = env.deepcopy(orig_value)
        end
        setmetatable(copy, env.deepcopy(getmetatable(orig)))
    else -- number, string, boolean, etc
        copy = orig
    end
    return copy
end

function env.pad(n)
    local s = ""

    for i = 1, n do
        s = s.."     "
    end

    return s
end

function env.dump(t, depth)
    if DevTools_Dump then
        DevTools_Dump(t)
        return
    end

    if type(t) == "table" then
        if not depth then
            depth = 1
            debug("Table dump: {")
        end

        if depth > 5 then return end

        for k,v in pairs(t) do
            if type(v) == "string" or type(v) == "number" then
                print(pad(depth)..k.."="..v)
            elseif type(v) == "table" then
                print(pad(depth)..k.."={")
                env.dump(v, depth+1)
                print(pad(depth).."}")
            end
        end

        print("}")
    else
        print(t)
    end
end

-- Copies values from t2 to t1, recursively
-- With an option to only overwrite nils
function env.updateTable(t1, t2, onlyNilValues)
    for k, v in pairs(t2) do
        if type(v) == "table" then
            if t1[k] == nil then t1[k] = {} end
            env.updateTable(t1[k], v)
        else
            if not onlyNilValues or t1[k] == nil then
                t1[k] = v
            end
        end
    end
end


Options.lua
local _, env = ...

-- Create options frame
local optionsFrame = CreateFrame("Frame", nil, UIParent)
optionsFrame.name = "Instance History Extra"
optionsFrame.settingChanges = {}
function optionsFrame.default()
    local config = InstanceHistoryExtraSV.CONFIG
    config = env.deepcopy(env.configDefaults)
    optionsFrame.populateOptions()
    env.f.drawProgressBar()
end
env.optionsFrame = optionsFrame
InterfaceOptions_AddCategory(optionsFrame)

local scale = UIParent:GetEffectiveScale()
local swh, sh = floor(GetScreenWidth() * scale / 2), floor(GetScreenHeight() * scale)
local elements = {
    {type="check", text="Debug Mode", key="debugMode",
        desc="Prints debug messages to the default chat frame."},
    {type="check", text="Report Resets", key="reportResets",
        desc="Reports instance resets to group channels."},
    {type="check", text="Report Time Only When Locked", key="reportLockedOnly",
        desc="When disabled, will always report time until oldest instance expires, unless under hourly lock. Otherwise, only report time when locked out." },
    {type="check", text="Colorize Display Segment", key="colorProgress",
        desc="Applies color to individual segments of the display."},
    {type="check", text="Force 24-Hour Display", key="force24H",
        desc="Forces display to show 24 hours instead of dynamically changing."},
    {type="padding"},
    {type="slider", text="Display Update Interval", key="updateInterval", min=0.5, max=10, step=0.5,
        desc="Number of seconds to wait between each display update."},
    {type="slider", text="Display Threshold", key="displayMin", min=1, max=30, step=1,
        desc="Minimum number of instances before display appears."},
    {type="slider", text="Width", key="width", min=0, max=500, step=1},
    {type="slider", text="Height", key="height", min=0, max=100, step=1},
    {type="slider", text="Horizontal Position", key="xOffset", min=-swh, max=swh, step=1},
    {type="slider", text="Vertical Position", key="yOffset", min=0, max=sh, step=1},
    {type="slider", text="Font Size", key="fontSize", min=8, max=32, step=1},
}
local xSpace = 290
local xOfs = 20

-- Header
local fs = optionsFrame:CreateFontString(nil, "ARTWORK")
fs:SetFontObject(GameFontNormalLarge)
fs:SetText("Instance History Extra")
fs:SetPoint("TOPLEFT", optionsFrame, "TOPLEFT", 20, -20)

local line = optionsFrame:CreateLine()
line:SetColorTexture(1, 1, 1, 0.3)
line:SetStartPoint("TOPLEFT", fs, 0, -3 - fs:GetStringHeight())
line:SetEndPoint("TOPRIGHT", fs, 5, -3 - fs:GetStringHeight())
line:SetThickness(1.5)

-- Populate options
optionsFrame.formElements = {}
local ypos = -20
for i, d in ipairs(elements) do
    local e
    local x = (i - 1) % 2

    if d.type == "check" then
        if x == 0 then
            ypos = ypos - 30
        end
        e = CreateFrame("CheckButton", nil, optionsFrame, "UICheckButtonTemplate")
        e.text:SetText(d.text)
        e.text:SetFontObject(GameFontHighlight)
        e:SetPoint("TOPLEFT", optionsFrame, "TOPLEFT", xOfs + x * xSpace, ypos)
        e:SetScript("OnClick", function(self)
            local config = InstanceHistoryExtraSV.CONFIG
            config[self.configKey] = self:GetChecked()
            env.f.drawProgressBar()
        end)
        if d.desc then
            e:SetScript("OnEnter", function(self)
                GameTooltip:SetOwner(self, "ANCHOR_BOTTOMRIGHT")
                GameTooltip:AddLine(d.text)
                GameTooltip:AddLine(d.desc, 1, 1, 1, true)
                GameTooltip:Show()
            end)
            e:SetScript("OnLeave", function()
                GameTooltip:Hide()
            end)
        end
    elseif d.type == "slider" then
        if x == 0 then
            ypos = ypos - 60
        end
        -- Create slider
        e = CreateFrame("Slider", nil, optionsFrame, "OptionsSliderTemplate")
        e:SetOrientation("HORIZONTAL")
        e:SetWidth(xSpace - 50)
        e:SetMinMaxValues(d.min, d.max)
        e:SetValueStep(d.step)
        e:SetObeyStepOnDrag(true)
        e:SetPoint("TOPLEFT", optionsFrame, "TOPLEFT", xOfs + x * xSpace + 25, ypos)
        e:SetScript("OnValueChanged", function(self, val)
            self.editBox:SetNumber(val)
            -- save config, refresh progress bar
            local config = InstanceHistoryExtraSV.CONFIG
            config[self.configKey] = val
            env.f.drawProgressBar()
        end)
        -- Create attached editbox
        e.editBox = CreateFrame("EditBox", nil, e)
        e.editBox:SetBackdrop({
            bgFile = "Interface\\DialogFrame\\UI-DialogBox-Background",
            edgeFile = "Interface\\PVPFrame\\UI-Character-PVP-Highlight",
            edgeSize = 16,
            insets = {left = 8, right = 6, top = 8, bottom = 8},
        })
        e.editBox:SetBackdropColor(0, 0, 0, 1)
        e.editBox:SetBackdropBorderColor(0.5, 0.5, 0.5, 1)
        e.editBox:SetSize(60, 30)
        e.editBox:SetMultiLine(false)
        e.editBox:SetAutoFocus(false)
        e.editBox:SetJustifyH("CENTER")
        e.editBox:SetJustifyV("CENTER")
        e.editBox:SetFontObject(GameFontWhiteSmall)
        e.editBox:SetScript("OnArrowPressed", function(self, key)
            local vmin, vmax = self:GetParent():GetMinMaxValues()
            local step = self:GetParent():GetValueStep()
            local val = self:GetNumber()
            -- increase/decrease value depending on step value
            if key == "UP" then
                val = val + step
            elseif key == "DOWN" then
                val = val - step
            end
            -- clamping
            val = max(vmin, min(vmax, val))
            -- set numbers (thanks floats)
            self:GetParent():SetValue(val)
            self:SetNumber(self:GetParent():GetValue(val))
            -- save config, refresh progress bar
            local config = InstanceHistoryExtraSV.CONFIG
            config[self:GetParent().configKey] = self:GetNumber()
            env.f.drawProgressBar()
        end)
        e.editBox:SetScript("OnEnterPressed", function(self)
            local vmin, vmax = self:GetParent():GetMinMaxValues()
            local val = self:GetNumber()
            -- clamping
            val = max(vmin, min(vmax, val))
            self:SetNumber(val)
            self:GetParent():SetValue(val)
            -- to accommodate step values
            self:SetNumber(self:GetParent():GetValue(val))
            self:ClearFocus()
            -- save config, refresh progress bar
            local config = InstanceHistoryExtraSV.CONFIG
            config[self:GetParent().configKey] = self:GetNumber()
            env.f.drawProgressBar()
        end)
        e.editBox:SetScript("OnEscapePressed", function(self)
            self:SetText(self:GetParent():GetValue(val))
            self:ClearFocus()
        end)
        if d.desc then
            e:SetScript("OnEnter", function(self)
                GameTooltip:SetOwner(self, "ANCHOR_BOTTOMRIGHT")
                GameTooltip:AddLine(d.text)
                GameTooltip:AddLine(d.desc, 1, 1, 1, true)
                GameTooltip:Show()
            end)
            e:SetScript("OnLeave", function()
                GameTooltip:Hide()
            end)
        end
        -- Modify labels
        e.Low:SetText(d.min)
        e.High:SetText(d.max)
        e.Text:SetText(d.text)
        e.Text:SetFontObject(GameFontNormal)
        e.editBox:SetPoint("TOP", e.Text, "BOTTOM", 0, -10)
    end

    if e then
        e.configKey = d.key
        optionsFrame.formElements[d.key] = e
    end
end

-- Author info
local text = optionsFrame:CreateFontString(nil, "ARTWORK")
text:SetFontObject(GameFontDisableSmall)
text:SetText(env.c.authorInfo)
text:SetPoint("BOTTOMRIGHT", optionsFrame, "BOTTOMRIGHT", -10, 10)

function optionsFrame.populateOptions()
    local config = InstanceHistoryExtraSV.CONFIG
    for k, v in pairs(config) do
        local e = optionsFrame.formElements[k]
        if e then
            local type = e:GetObjectType()
            if type == "CheckButton" then
                e:SetChecked(v)
            elseif type == "Slider" then
                e:SetValue(v)
                e.editBox:SetNumber(v)
                e.editBox:SetCursorPosition(0)
            end
        end
    end
end


Instances.lua
local _, env = ...

local f = {}

function f.InGroup()
    if IsInRaid() then return "RAID"
    elseif GetNumGroupMembers() > 0 then return "PARTY"
    else return nil end
end

function f.histZoneKey()
    local instname, insttype, diff, diffname, maxPlayers, playerDifficulty, isDynamicInstance = GetInstanceInfo()

    -- PVP doesn't count
    if insttype == nil or insttype == "none" or insttype == "arena" or insttype == "pvp" then
        return nil
    end

    -- 40-man raids don't count
    if insttype == "raid" and maxPlayers == 40 then
        return nil
    end

    -- LFG instances don't count, but Holiday Event counts
    --[[
    if (IsInLFGDungeon() or IsInScenarioGroup()) and diff ~= 19 then
        return nil
    end  ]]

    local db = InstanceHistoryExtraSV[env.c.thisToon]

    -- Check if we're locked (using FindInstance so we don't complain about unsaved unknown instances)
    local truename = f.findInstance(instname, insttype == "raid")
    local locked = false
    local inst = truename and db.Instances[truename]
    inst = inst and inst[env.c.thisToon]

    for d = 1, env.c.maxdiff do
        if inst and inst[d] and inst[d].Locked then
            locked = true
        end
    end

    if diff == 1 and maxPlayers == 5 then -- never locked to 5-man regs
        locked = false
    end

    local toonstr = env.c.thisToon

    local desc = toonstr .. ": " .. instname

    if diffname and #diffname > 0 then
        desc = desc .. " - " .. diffname
    end

    local key = toonstr..":"..instname..":"..insttype..":"..diff

    if not locked then
        key = key..":"..db.histGeneration
    end

    return key, desc, locked
end

function f.normalizeName(str)
    return str:gsub("%p",""):gsub("%s"," "):gsub("%s%s"," "):gsub("^%s+",""):gsub("%s+$",""):upper()
end

-- some instances (like sethekk halls) are named differently by GetSavedInstanceInfo() and LFGGetDungeonInfoByID()
-- we use the latter name to key our database, and this function to convert as needed
function f.findInstance(name, raid)
    if not name or #name == 0 then return nil end

    local db = InstanceHistoryExtraSV[env.c.thisToon]

    local nname = f.normalizeName(name)
    -- first pass, direct match
    local info = db.Instances[name]

    if info then
        return name, info.LFDID
    end

    -- hyperlink id lookup: must precede substring match for ticket 99
    -- (so transInstance can override incorrect substring matches)
    for i = 1, GetNumSavedInstances() do
        local link = GetSavedInstanceChatLink(i) or ""
        local lid,lname = link:match(":(%d+):%d+:%d+\124h%[(.+)%]\124h")
        lname = lname and f.normalizeName(lname)
        lid = lid and tonumber(lid)
        local lfdid = lid and env.c.transInstance[lid]
        if lname == nname and lfdid then
            -- zero idea what was `addon` referring to
            local truename = addon:UpdateInstance(lfdid)
            if truename then
                return truename, lfdid
            end
        end
    end
    -- normalized substring match
    for truename, info in pairs(db.Instances) do
        local tname = f.normalizeName(truename)
        if (tname:find(nname, 1, true) or nname:find(tname, 1, true)) and
                info.Raid == raid then -- Tempest Keep: The Botanica
            --env.debug("FindInstance("..name..") => "..truename)
            return truename, info.LFDID
        end
    end
    return nil
end

function f.generationAdvance()
    local db = InstanceHistoryExtraSV[env.c.thisToon]
    env.debug("HistoryUpdate generation advance")
    db.histGeneration = (db.histGeneration + 1) % 100000
    db.sess.delayedReset = false
end

function f.HistoryUpdate(forcereset, forcemesg)
    local db = InstanceHistoryExtraSV[env.c.thisToon]
    local config = InstanceHistoryExtraSV.CONFIG
    db.histGeneration = db.histGeneration or 1

    if forcereset and f.histZoneKey() then -- delay reset until we zone out
        env.debug("HistoryUpdate reset delayed")
        db.sess.delayedReset = true
    end

    if (forcereset or db.sess.delayedReset) and not f.histZoneKey() then
        f.generationAdvance()
    elseif db.lastLoc then
        if not db.sess.enterLoc then
            -- delay until enterLoc is defined
            C_Timer.After(1, f.HistoryUpdate)
            return
        end

        if db.sess.enterLoc.instance == db.lastLoc.instance and db.lastLoc.instance >= 0 and db.sess.enterLoc.subzone ~= db.lastLoc.subzone then
            env.debug("Offline forced reset detected.")
            f.generationAdvance()
        else
            env.debug("Location data resolved, no reset detected.")

            --[[if db.config.debugMode then
                print(db.sess.enterLoc.instance, db.sess.enterLoc.subzone, db.lastLoc.instance, db.lastLoc.subzone)
            end]]
        end

        --[[if db.lastLoc then
            env.debug("lastLoc cleared")
        end]]

        db.lastLoc = nil
    end

    local now = time()

    if db.sess.delayUpdate and now < db.sess.delayUpdate then
        --env.debug("HistoryUpdate delayed")
        C_Timer.After(db.sess.delayUpdate - now + 0.05, f.HistoryUpdate)
        return
    end

    local newzone, newdesc, locked = f.histZoneKey()

    -- touch zone we left
    if db.sess.histLastZone then
        local lz = db.History[db.sess.histLastZone]
        if lz then
            lz.last = now
        end
    end

    db.sess.histLastZone = newzone
    db.sess.histInGroup = f.InGroup()

    -- touch/create new zone
    if newzone then
        local nz = db.History[newzone]

        -- if last touched at least 30 mins ago, it's a new instance
        if nz and now - nz.last >= 1800 then
            f.generationAdvance()
            newzone, newdesc, locked = f.histZoneKey()
            db.sess.histLastZone = newzone

            nz = nil
        end

        if not nz then
            nz = { create = now, desc = newdesc }
            db.History[newzone] = nz

            if locked then -- creating a locked instance, delete unlocked version
                db.History[newzone..":"..db.histGeneration] = nil
            end
        end

        nz.last = now
    end

    -- reap old zones
    local livecnt = 0
    local oldestkey, oldesttime

    for zk, zi in pairs(db.History) do
        if now > zi.last + env.c.histReapTime then
            env.debug("Reaping %s",zi.desc)
            db.History[zk] = nil
            db.colorOffset = (db.colorOffset + 1) % #env.c.colors
        else
            livecnt = livecnt + 1

            if not oldesttime or zi.last < oldesttime then
                oldestkey = zk
                oldesttime = zi.last
            end
        end
    end

    local oldestrem = oldesttime and (oldesttime + env.c.histReapTime - now)
    local oldestremt = oldestrem and SecondsToTime(oldestrem,false,false,1):lower() or "n/a"

    if config.debugMode then
        local msg = livecnt.." live instances, oldest ("..(oldestkey or "none")..") expires in "..oldestremt..". Current Zone="..(newzone or "nil")
        if msg ~= db.sess.lasthistdbg then
            db.sess.lasthistdbg = msg
            env.debug(msg)
        end
    end

    db.sess.histLiveCount = livecnt
    db.sess.histOldest = oldestremt
end

-- fixme localize or something
function f.doExplicitReset(instancemsg, failed)
    if f.InGroup() and not UnitIsGroupLeader("player") then
        return
    end

    local config = InstanceHistoryExtraSV.CONFIG

    if not failed then
        f.HistoryUpdate(true)
    end

    local reportchan = f.InGroup()

    if reportchan then
        if not failed then
            C_ChatInfo.SendAddonMessage(env.c.prefix, "GENERATION_ADVANCE", reportchan)
        end
        if config.reportResets then
            local msg = instancemsg or RESET_INSTANCES
            msg = msg:gsub("\1241.+.+","") -- ticket 76, remove |1;; escapes on koKR
            SendChatMessage("All instances have been reset.", reportchan)
        end
    end
end

hooksecurefunc("ResetInstances", f.doExplicitReset)

function f.zoneChanged(extraDelay)
    -- delay updates while settings stabilize
    local db = InstanceHistoryExtraSV[env.c.thisToon]

    local waittime = 3 + math.max(0,10 - GetFramerate()) + (extraDelay or 0)
    local d = time() + waittime

    if d > (db.sess.delayUpdate or 0) then
        db.sess.delayUpdate = d
    end

    C_Timer.After(waittime + 0.05, f.HistoryUpdate)
end

function f.getLocation()
    local _, instanceType = GetInstanceInfo()
    local instanceID = select(8, GetInstanceInfo())

    local loc

    if instanceType == "none" then
        loc = {}
        loc.instance = -1
    else
        loc = {}
        loc.instance = instanceID
        loc.subzone = GetSubZoneText()
    end

    return loc
end

function f.doAutoReset()
    if not env.autoReset then
        return
    end

    if not UnitIsGroupLeader("player") then
        env.debug("Player is not leader, auto reset disabled.")
        env.autoReset = nil
        return
    end

    local name = env.autoReset

    if UnitExists(name) then
        if not UnitIsConnected(name) then
            C_Timer.After(1, function()
                ResetInstances()
            end)
            env.autoReset = nil
            return
        end
    else
        env.debug("Sender does not exist, auto reset disabled.")
        env.autoReset = nil
        return
    end
end

function f.updateProgress(noCallback)
    if not noCallback then
        -- Originally this function runs every frame, but this is close enough
        C_Timer.After(0.1, f.updateProgress)
    end

    if env.autoReset then
        f.doAutoReset()
    end

    local db = InstanceHistoryExtraSV[env.c.thisToon]
    local config = InstanceHistoryExtraSV.CONFIG

    if GetTime() - db.lastDisplayUpdate >= config.updateInterval then
        local s = {}

        local count, daycount = 0, 0
        local now = time()

        for _, v in pairs(db.History) do
            if now - 3600 <= v.last and v.last <= now then
                count = count + 1
                daycount = daycount + 1
            elseif now - 24*3600 <= v.last and v.last <= now then
                daycount = daycount + 1
            end
        end

        s.show = daycount >= config.displayMin

        if count >= config.displayMin and not config.force24H then
            s.total = 3600
        else
            s.total = 24*3600
        end

        -- Only do a toggle if we won't make an empty progress bar
        if env.progressBar.toggle and count > 0 then
            s.total = s.total == 3600 and 24*3600 or 3600
        end

        s.additionalProgress = {}

        local start = now - s.total
        local czk = f.histZoneKey()
        local ordered = {}

        for k,v in pairs(db.History) do
            if k == czk and now - v.last < 1800 then
                v.last = now
            end

            table.insert(ordered, {k, v.create})
        end

        table.sort(ordered, function(a,b) return a[2]<b[2] end)

        for _, t in pairs(ordered) do
            local k = t[1]
            local v = db.History[k]

            if v.last >= start then
                local o = {
                    max = max(0, min(s.total, v.last - start)),
                    min = max(0, min(s.total, v.create - start))
                }

                table.insert(s.additionalProgress, o)
            end
        end

        --DevTools_Dump(s.additionalProgress)
        db.lastDisplayUpdate = GetTime()

        f.displayProgress(s, daycount)
    end
end

function f.displayProgress(s, n)
    local p = env.progressBar
    local width, height = p:GetSize()
    local db = InstanceHistoryExtraSV[env.c.thisToon]
    local config = InstanceHistoryExtraSV.CONFIG
    local cmap = env.c.colors

    if not s.show then
        p:Hide()
    else
        p:Show()
        -- Add additional textures if we need more
        while #p.textures < #s.additionalProgress do
            local t = p:CreateTexture()
            t:SetTexture(env.c.texture)
            table.insert(p.textures, t)
        end

        -- Position and align textures correctly, and hide excess textures
        for i, t in ipairs(p.textures) do
            local chunk = s.additionalProgress[i]
            if chunk then
                local chunk_start = max(0, chunk.min / s.total)
                local chunk_end = min(1, chunk.max / s.total)
                local chunk_length = (chunk.max - chunk.min) / s.total

                if chunk_length > 0 then
                    t:SetSize(chunk_length * width, height)
                    t:SetPoint("TOPLEFT", p, "TOPLEFT", chunk_start * width, 0)

                    if config.colorProgress then
                        local c = (i - 1 + (n - #s.additionalProgress) + db.colorOffset) % #cmap
                        t:SetVertexColor(unpack(cmap[c + 1]))
                    else
                        t:SetVertexColor(1, 1, 1)
                    end

                    t:SetTexCoord(chunk_start, chunk_end, 0, 1)
                    t:Show()
                else
                    t:Hide()
                end
            else
                t:Hide()
            end
        end
    end
end

function f.updateText(noCallback)
    if not noCallback then
        C_Timer.After(1, f.updateText)
    end

    local now = time()
    local db = InstanceHistoryExtraSV[env.c.thisToon]
    local config = InstanceHistoryExtraSV.CONFIG
    local fs = env.progressBar.text

    local oldestTime, oldestDayTime
    local count = 0
    local daycount = 0

    for _, v in pairs(db.History) do
        if now - 3600 <= v.last then
            count = count + 1
            daycount = daycount + 1

            if not oldestTime or v.last < oldestTime then
                oldestTime = v.last
            end
            if not oldestDayTime or v.last < oldestDayTime then
                oldestDayTime = v.last
            end
        elseif now - 24*3600 <= v.last then
            daycount = daycount + 1

            if not oldestDayTime or v.last < oldestDayTime then
                oldestDayTime = v.last
            end
        end
    end

    local rem = oldestTime and (oldestTime + 3600 - now)
    local remday = oldestDayTime and (oldestDayTime + 24*3600 - now)

    local instanceStr, timestr
    local plural = {[true]='instances', [false]='instance'}

    if daycount > 0 then
        if count > 0 then
            instanceStr = string.format("You have entered %d %s in the last 24 hours,\nand %d %s in the past hour.", daycount, plural[daycount>1], count, plural[count>1])
        else
            instanceStr = string.format("You have entered %d %s in the last 24 hours.", daycount, plural[daycount>1])
        end
    end

    if config.reportLockedOnly then
        if daycount == env.c.histLimit and count == env.c.histLimitHourly then
            timestr = (rem and remday) and SecondsToTime(max(rem, remday)):lower() or "n/a"
        elseif daycount == env.c.histLimit then
            timestr = remday and SecondsToTime(remday):lower() or "n/a"
        elseif count == env.c.histLimitHourly then
            timestr = rem and SecondsToTime(rem):lower() or "n/a"
        end
    else
        if count == env.c.histLimitHourly then
            if daycount == env.c.histLimit then
                timestr = (rem and remday) and SecondsToTime(max(rem, remday)):lower() or "n/a"
            else
                timestr = rem and SecondsToTime(rem):lower() or "n/a"
            end
        elseif daycount > 0 then
            timestr = remday and SecondsToTime(remday):lower() or "n/a"
        end
    end
    
    timestr = (timestr == "") and "a jiffy" or timestr

    if timestr then
        instanceStr = instanceStr .. string.format("\nMore instances will be available in %s.", timestr)
    end

    local currentFontSize = select(2, fs:GetFont())
    if currentFontSize ~= floor(config.fontSize + 0.5) then
        fs:SetFont(env.c.font, config.fontSize, "OUTLINE")
    end
    fs:SetText(instanceStr)
    fs:SetPoint("BOTTOM", env.progressBar, "TOP", 0, 4)
end

function f.drawProgressBar()
    local db = InstanceHistoryExtraSV[env.c.thisToon]
    local config = InstanceHistoryExtraSV.CONFIG
    local p = env.progressBar
    local scale = UIParent:GetEffectiveScale()

    p:SetSize(config.width / scale, config.height / scale)
    p:SetPoint("TOP", UIParent, "TOP", config.xOffset / scale, -config.yOffset / scale)

    -- redraw the sub-progress bars
    db.lastDisplayUpdate = 0
    f.updateProgress(true)
    f.updateText(true)
end

env.f = f
